src/core/Events.js

```diff
@@ -152,30 +145,24 @@ export var Events = {
 			return;
 		}
 
-		if (context === this) {
-			context = undefined;
-		}
-
 		if (typeof fn !== 'function') {
 			console.warn('wrong listener type: ' + typeof fn);
 			return;
 		}
+
 		// find fn and remove it
-		for (i = 0, len = listeners.length; i < len; i++) {
-			var l = listeners[i];
-			if (l.ctx !== context) { continue; }
-			if (l.fn === fn) {
-				if (this._firingCount) {
-					// set the removed listener to noop so that's not called if remove happens in fire
-					l.fn = Util.falseFn;
-
-					/* copy array in case events are being fired */
-					this._events[type] = listeners = listeners.slice();
-				}
-				listeners.splice(i, 1);
+		var listener = this._listens(type, fn, context);
+		if (listener) {
+			if (this._firingCount) {
+				// set the removed listener to noop so that's not called if remove happens in fire
+				listener.fn = Util.falseFn;
 
-				return;
+				/* copy array in case events are being fired */
+				this._events[type] = listeners = listeners.slice();
 			}
+			var idx = listeners.indexOf(listener);

```

Comment:
I just realized that this is a perf regression for `off`, since we now essentially have to loop over the listeners twice — first to find the listener in `_listen`, and then to fined the index for removal. 

One way to fix this would be to return the index rather than the listener in `_listen`, but then we have to be careful in `if` conditions elsewhere, changing from `if (this._listens(...))` to `if (this._listens(...) !== false)`.

Reply:
I fixed that with returning a object:
```
return {
	index: i,
	listener: listeners[i]
};
```